cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/DefaultSystemAppTaskContext.java

```diff
@@ -0,0 +1,182 @@
+    this.artifactManager = artifactManagerFactory.create(new NamespaceId(artifactNameSpace), retryStrategy);
+    this.closeableList = new ArrayList<>();
+    File pluginsDir = createTempFolder();
+    instantiator = new PluginInstantiator(cConf, artifactClassLoader, pluginsDir);
+    closeableList.add(() -> {
+      try {
+        instantiator.close();
+      } finally {
+        DirUtils.deleteDirectoryContents(pluginsDir, true);
+      }
+    });
+    protoArtifactId = new io.cdap.cdap.proto.id.ArtifactId(artifactNameSpace, artifactId.getName(),
+                                                           artifactId.getVersion().getVersion());
+  }
+
+  private File createTempFolder() {
+    File tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    try {
+      return Files.createTempDirectory(tmpDir.toPath(), "plugins").toFile();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Map<String, String> getPreferencesForNamespace(String namespace, boolean resolved) throws IOException {
+    try {
+      return preferencesFetcher.get(new NamespaceId(namespace), resolved).getProperties();
+    } catch (NotFoundException nfe) {
+      throw new IllegalArgumentException(String.format("Namespace '%s' does not exist", namespace), nfe);
+    }
+  }
+
+  @Override
+  public PluginConfigurer createPluginConfigurer(String namespace) throws IOException {
+    return new DefaultPluginConfigurer(protoArtifactId, new NamespaceId(namespace), instantiator, pluginFinder);
+  }
+
+  @Override
+  public Map<String, String> evaluateMacros(String namespace, Map<String, String> macros, MacroEvaluator evaluator,
+                                            MacroParserOptions options) throws InvalidMacroException {
+    MacroParser macroParser = new MacroParser(evaluator, options);
+    Map<String, String> evaluated = new HashMap<>();
+
+    for (Map.Entry<String, String> property : macros.entrySet()) {
+      String key = property.getKey();
+      String val = property.getValue();
+      evaluated.put(key, macroParser.parse(val));
+    }
+    return evaluated;
+  }
+
+  @Override
+  public ArtifactManager getArtifactManager() {
+    return artifactManager;
+  }
+
+  @Override
+  public String getServiceName() {
+    return serviceName;
+  }
+
+  @Override
+  protected DiscoveryServiceClient getDiscoveryServiceClient() {
+    return discoveryServiceClient;
+  }
+
+  @Override
+  public List<SecureStoreMetadata> list(String namespace) throws Exception {
+    return Retries.callWithRetries(() -> secureStore.list(namespace), retryStrategy);
+  }
+
+  @Override
+  public SecureStoreData get(String namespace, String name) throws Exception {
+    return Retries.callWithRetries(() -> secureStore.get(namespace, name), retryStrategy);
+  }
+
+  @Override
+  public void close() {
+    for (Closeable closeable : closeableList) {
```

Comment:
Seems like only the `instantiator` needs to be closed. No need to use a list for that.

Reply:
Done.